File: tests/fixtures/m3_ops.py
# M3 管理操作双语言一致性矩阵（a1_matrix 驱动）：E 码表对账、金额显示值
# 换算、borrow / repay / buy 关键码路径与资金线。与 m3_ops.js 逐行对照，
# 日志序列与世界终态哈希必须一致。
#
# 世界前提（demo_world）：12×8 无墙、1 机器人、1 货架 (6,3)、1 充电桩
# (2,5)、1 装卸位、金币 200、无欠款。

import math as _m


def loop():
    # E 码表对账：键集合排序后逐字一致（codes::ALL ↔ 两份 bootstrap）。
    Game.log("e:" + ",".join(sorted(Game.E.keys())))

    # 金额线统一以 milli 记录（floor(x*1000+0.5)，与绑定层换算同式）。
    def milli(x):
        return _m.floor(x * 1000 + 0.5)

    def funds():
        Game.log("f:" + str(milli(Game.gold)) + "," + str(milli(Game.debt)))

    # 借款：基本路径、非正数 / 非数值 / 超安全范围拒绝、超额拒绝。
    Game.log("b1:" + Game.borrow(50.5))
    funds()
    Game.log("b2:" + Game.borrow(0.1))  # 浮点陷阱值：0.1*1000 两侧同误差
    Game.log("d:" + str(milli(Game.debt)))
    Game.log("b3:" + Game.borrow(0))
    Game.log("b4:" + Game.borrow(-1))
    Game.log("b5:" + Game.borrow("x"))
    Game.log("b6:" + Game.borrow(1e30))
    Game.log("b7:" + Game.borrow(2000))  # 超信用额度

    # 还款：部分归还、超额钳定到欠款。
    Game.log("r1:" + Game.repay(0.1))
    Game.log("r2:" + Game.repay(100000))
    funds()

    # 商店：成功路径与错误码矩阵（放置 → 资金次序）。
    Game.log("y1:" + Game.buy("shelf", 5, 5))
    Game.log("sh:" + str(len(Game.shelves())) + "," + str(Game.shelves()[0].capacity))
    Game.log("y2:" + Game.buy("robot", 6, 6))  # 余额 24.9 < 650
    Game.log("b8:" + Game.borrow(700))  # 借款解围
    Game.log("y3:" + Game.buy("robot", 6, 6))
    Game.log("rb:" + str(len(Game.robots())) + "," + str(Game.robots()[1].energy))
    Game.log("y4:" + Game.buy("dock", 0, 0))  # 边界格但无墙 → NOT_ON_WALL
    Game.log("y5:" + Game.buy("crane", 1, 1))  # 未知 kind
    Game.log("y6:" + Game.buy("charger", 2, 5))  # 既有充电桩占格
    Game.log("y7:" + Game.buy("shelf", -1, 5))  # 越界
    Game.log("y9:" + Game.buy("charger", 7, 7))  # 余额 74.9 < 300
    Game.log("b9:" + Game.borrow(300))  # 恰好补到信用额度
    Game.log("b10:" + Game.borrow(1))  # 额度已满
    Game.log("y10:" + Game.buy("charger", 7, 7))
    Game.log("ch:" + str(len(Game.chargers())))
    Game.log("r3:" + Game.repay(50))
    Game.log("x1:" + Game.destroy(Game.shelves()[1].id))  # 销毁新购货架（半价退款）
    Game.log("sh:" + str(len(Game.shelves())))
    funds()

File: crates/runtime/tests/m3_ops.rs
//! M3 管理操作贯通 JS 宿主（docs/architecture/03 管理操作当 tick 可见）：
//! buy / borrow / repay 经绑定层往返，镜像增量回放后同 tick 查询立即可见
//! （快照纯度的唯一例外）；初始化阶段照拒 INIT_PHASE；计息随 Session 闭环。

mod common;

use common::{demo_world, fixture};
use ztw_api::harness::{OutcomeKind, Session, SessionConfig};

/// 增量回放后同 tick 可见：buy 的对象 / 资金线、borrow / repay 的欠款线
/// 不等下一次全量镜像（与 a1_matrix 的双语言对账共用 m3_ops.js 序列）。
#[test]
fn manage_ops_visible_same_tick_via_delta_replay() {
    let mut s = Session::new(SessionConfig::new(common::host_bin()), demo_world());
    assert!(
        s.load_code(&fixture("m3_ops.js")).ok,
        "初始化失败：{:?}",
        s.fault
    );
    let out = s.tick();
    assert_eq!(out.kind, OutcomeKind::Ok, "{:?}", s.fault);
    // 终态：原货架留存（新购的已销毁）、充电桩 +1、机器人 +1；金币
    // 112.5 gold；欠款 950 gold + 首 tick 利息 0.95。
    assert_eq!(s.world.shelves.len(), 1);
    assert_eq!(s.world.chargers.len(), 2);
    assert_eq!(s.world.robots.len(), 2);
    assert_eq!(s.world.gold_milli, 112_500);
    assert_eq!(s.world.debt_milli, 950_950);
    // 日志锚点：NOT_ON_WALL 经绑定层原样返回（E 表同步的证据点）。
    assert!(
        s.logs.iter().any(|(_, l)| l == "y4:NOT_ON_WALL"),
        "{:?}",
        s.logs
    );
}

/// 初始化阶段禁用管理操作：返回 INIT_PHASE 码（不抛错、初始化照常成功），
/// 世界资金不动。
#[test]
fn manage_ops_rejected_in_init_phase() {
    let mut s = Session::new(SessionConfig::new(common::host_bin()), demo_world());
    let init = s
        .load_code(
            r#"
const seen = [];
seen.push(Game.borrow(10));
seen.push(Game.buy("shelf", 5, 5));
Game.log("init:" + seen.join(","));
function loop() {}
"#,
        )
        .ok;
    assert!(init, "初始化应成功（拒绝是返回码，不是异常）");
    let out = s.tick();
    assert_eq!(out.kind, OutcomeKind::Ok);
    assert!(
        s.logs
            .iter()
            .any(|(_, l)| l == "init:INIT_PHASE,INIT_PHASE"),
        "{:?}",
        s.logs
    );
    assert_eq!(s.world.gold_milli, 200_000, "初始化期资金不动");
    assert_eq!(s.world.debt_milli, 0);
    assert_eq!(s.world.shelves.len(), 1, "原货架仍在，未新增");
}

/// 计息经完整 Session 闭环：阶段 5 收尾按欠款 × 利率复利。
#[test]
fn interest_accrues_through_session_tick() {
    let mut s = Session::new(SessionConfig::new(common::host_bin()), demo_world());
    assert!(
        s.load_code("function loop() { if (Game.tick === 0) Game.borrow(100); }")
            .ok,
        "初始化失败：{:?}",
        s.fault
    );
    assert_eq!(s.tick().kind, OutcomeKind::Ok, "{:?}", s.fault);
    assert_eq!(s.world.debt_milli, 100_100, "100_000 + 首笔利息 100");
    assert_eq!(s.tick().kind, OutcomeKind::Ok, "{:?}", s.fault);
    assert_eq!(s.world.debt_milli, 100_200, "复利：100_100 + 100");
}

/// 增量回放失败不丢当 tick 可见性：borrow / buy 的资金线与新对象经
/// mirror.fetch 整体重建后立即可读（skip_delta_replay 注入；a0_boundary
/// 的 take 先例扩展到管理操作）。
#[test]
fn funds_and_buy_visible_after_replay_failure() {
    let cfg = SessionConfig::new(common::host_bin()).with_fault("skip_delta_replay");
    let mut s = Session::new(cfg, demo_world());
    assert!(
        s.load_code(
            r#"
function loop() {
  Game.borrow(50.5);
  Game.log("f:" + Math.floor(Game.gold * 1000 + 0.5) + "," + Math.floor(Game.debt * 1000 + 0.5));
  Game.buy("shelf", 5, 5);
  Game.log("s:" + Game.shelves().length + "," + Game.shelves()[1].id);
}
"#
        )
        .ok,
        "初始化失败：{:?}",
        s.fault
    );
    let out = s.tick();
    assert_eq!(out.kind, OutcomeKind::Ok, "{:?}", s.fault);
    assert!(
        s.logs.iter().any(|(_, l)| l == "f:250500,50500"),
        "{:?}",
        s.logs
    );
    assert!(
        s.logs.iter().any(|(_, l)| l.starts_with("s:2,")),
        "{:?}",
        s.logs
    );
    assert!(
        s.stats.last_exec.mirror_rebuilds >= 1,
        "应经 mirror.fetch 重建：{:?}",
        s.stats.last_exec
    );
    assert_eq!(s.world.shelves.len(), 2);
    assert_eq!(s.world.gold_milli, 75_500);
    assert_eq!(s.world.debt_milli, 50_550, "含首 tick 利息 50");
}
